### ISSUE TITLE
Better Typescript types

### ISSUE THREAD

**Issue by Linksku** [diegohaz/constate#117]
I'm learning TS and spent a long time figuring out how to type Constate properly, turns out it wasn't possible in TS 3, I needed TS 4:

```
declare module 'constate' {
  function constate<
    StoreProps,
    StoreValue,
    SplitValue extends ((p: StoreValue) => any),
    SplitValues extends ((p: StoreValue) => any)[],
  >(
    useValue: (p: StoreProps) => StoreValue,
    splitValue: SplitValue,
    ...splitValues: SplitValues
  ): [
    React.FC<StoreProps>,
    () => ReturnType<SplitValue>,
    ...(SplitValues extends infer A
    ? {
      [K in keyof A]: () => (A[K] extends (...args: any) => infer R ? R : never);
    }
    : never),
  ];

  function constate<
    StoreProps,
    StoreValue,
  >(useValue: (p: StoreProps) => StoreValue): [
    React.FC<StoreProps>,
    () => StoreValue,
  ];

  export default constate;
}
```

Since TS 4 is still in beta, this shouldn't be included in Constate yet, but I thought I'd share. Let me know if there's anything that can be improved!

**Comment by diegohaz** [comment]
The library is already written in TypeScript. Ideally, [you shouldn't do anything do get types working](https://codesandbox.io/s/github/diegohaz/constate/tree/master/examples/typescript?module=/App.tsx). Isn't it working for you?

**Comment by Linksku** [comment]
With the built-in types, it doesn't infer much:

```
const [CounterProvider, useCount] = constate(
  function useCounter() {
    const [count, setCount] = React.useState(0);
    return { count };
  },
  useCounter,
  value => value.count,
);

typeof useCount; // default: () => any, mine: () => number
```

**Comment by diegohaz** [comment]
@Linksku You're not using the function correctly. Please, take a look at the README again.

This is working as expected:
![Screenshot of typescript code with correct behavior](https://user-images.githubusercontent.com/3068563/90594896-f5fa9800-e1c1-11ea-84f2-d9329522ee9d.png)

**Comment by diegohaz** [comment]
But I agree that it will be much better with TS 4 spreads. And it'll hopefully fix #109.

**Comment by Linksku** [comment]
Hmm it does work after I remove my type, not sure what I did wrong initially, thanks!
